src/BeamAdapter/component/engine/WireRestShape.inl

```diff
@@ -182,120 +189,107 @@ void WireRestShape<DataTypes>::init()
+        densityList.clear();
 
-            if(d_straightLength.getValue()>= 0.001*this->d_length.getValue() )
-            {
-                int numNodes = (int) floor(5.0*d_straightLength.getValue() / d_length.getValue() );
-                densityList.push_back(numNodes);
-            }
-            if( d_straightLength.getValue() <=  0.999*d_length.getValue())
-            {
-                int numNodes = (int) floor(20.0*(1.0 - d_straightLength.getValue() / d_length.getValue()) );
-                densityList.push_back(numNodes);
-            }
+        for (unsigned int i = 0; i < keyPointList.size() - 1; ++i)
+        {
+            auto mat = l_sectionMaterials.get(i);
+            int nbrCollEdges = mat->getNbCollisionEdges();
+            densityList.push_back(nbrCollEdges);
         }
     }
 
-    if(!d_numEdgesCollis.getValue().size())
+    return true;
+}
+
+
+template <class DataTypes>
+bool WireRestShape<DataTypes>::initTopology()
+{
+    /// fill topology :
+    _topology->clear();
+    _topology->cleanup();
+
+    const type::vector<Real>& keyPts = d_keyPoints.getValue();
+    if (l_sectionMaterials.size() != keyPts.size() - 1)
     {
-        auto densityCol = sofa::helper::getWriteOnlyAccessor(d_numEdgesCollis);
-        densityCol.resize(keyPointList.size()-1);
-        for (unsigned int i=0; i<densityCol.size(); i++)
-            densityCol[i] = 20;
+        msg_error() << "Wrong number of inputs. Component can't be init. Number of input materials: " << l_sectionMaterials.size() << ", should be equal to keyPointList.size()-1. keyPointList.size() is equal to: " << keyPts.size();
+        return false;
     }
+    
+    Real prev_length = 0.0;
+    int prev_edges = 0;
+    int startPtId = 0; 
+    for (auto i = 0; i < l_sectionMaterials.size(); ++i)
+    {
+        // Add topology of the material 
+        int nbrVisuEdges = l_sectionMaterials.get(i)->getNbVisualEdges();
+        Real length = fabs(keyPts[i + 1] - keyPts[i]);
+        Real dx = length / nbrVisuEdges;
+
+        // add points from the material
+        for (int i = startPtId; i < nbrVisuEdges + 1; i++) {
+            _topology->addPoint(prev_length + i * dx, 0, 0);
+        }
 
-    msg_info() <<"WireRestShape end init" ;
-
-    // Prepare beam sections
-    double r 					= this->d_radius1.getValue();
-    double rInner 				= this->d_innerRadius1.getValue();
-    this->beamSection1._r 		= r;
-    this->beamSection1._rInner 	= rInner;
-    this->beamSection1._Iz		= M_PI*(r*r*r*r - rInner*rInner*rInner*rInner)/4.0;
-    this->beamSection1._Iy 		= this->beamSection1._Iz ;
-    this->beamSection1._J 		= this->beamSection1._Iz + this->beamSection1._Iy;
-    this->beamSection1._A 		= M_PI*(r*r - rInner*rInner);
-    this->beamSection1._Asy 	= 0.0;
-    this->beamSection1._Asz 	= 0.0;
+        // add segments from the material
+        for (int i = prev_edges; i < prev_edges + nbrVisuEdges; i++) {
+            _topology->addEdge(i, i + 1);
+        }
```

Comment:
I think we should clarify this point, as the topology is defined by the number of visual edges...
it is adding more confusion to the code.
This topology is only used by the visual part I suppose ? (aka the centerlines used for the edge2quadtopomapping)